src/fundus/scraping/html.py

```diff
@@ -132,68 +188,63 @@ class HTML:
 class HTMLSource:
     def __init__(
         self,
-        url_source: Iterable[str],
+        url_source: AsyncIterable[str],
         publisher: Optional[str],
-        url_filter: Optional[UrlFilter] = None,
-        max_threads: int = 10,
-        delay: Optional[Callable[[], float]] = None,
+        url_filter: Optional[URLFilter] = None,
         request_header: Optional[Dict[str, str]] = None,
     ):
         self.url_source = url_source
         self.publisher = publisher
-        self.url_filter = url_filter or (lambda url: not bool(url))
-        self.max_threads = max_threads
-        self.delay = delay
+        self.url_filter = [] if not url_filter else [url_filter]
         self.request_header = request_header or _default_header
         if isinstance(url_source, URLSource):
             url_source.set_header(self.request_header)
 
-    def _batched_fetch(self) -> Generator[List[Optional[HTML]], int, None]:
-        with requests.Session() as session:
+    def add_url_filter(self, url_filter: URLFilter):
+        self.url_filter.append(url_filter)
+
+    def _filter(self, url: str) -> bool:
+        for f in self.url_filter:
+            if f(url):
+                return True
+        return False
+
+    async def async_fetch(self) -> AsyncIterator[HTML]:
+        async for url in self.url_source:
+            if not validate_url(url):
```

Comment:
Its probably not best practice that there is filtering and validating of a given url. In my opinion, these two could be merged.